diff -Nrdu binutils-2.28/bfd/config.bfd binutils-2.28_cosmos/bfd/config.bfd
--- binutils-2.28/bfd/config.bfd	2017-03-02 09:23:53.000000000 +0100
+++ binutils-2.28_cosmos/bfd/config.bfd	2017-07-20 11:17:20.946009399 +0200
@@ -435,6 +435,10 @@
     targ_defvec=arm_elf32_le_vec
     targ_selvecs=arm_elf32_be_vec
     ;;
+  arm-*-cosmos*)
+    targ_defvec=arm_elf32_le_vec
+    targ_selvecs=arm_elf32_be_vec
+    ;;
   arm-*-rtems*)
     targ_defvec=arm_elf32_le_vec
     targ_selvecs=arm_elf32_be_vec
diff -Nrdu binutils-2.28/config.sub binutils-2.28_cosmos/config.sub
--- binutils-2.28/config.sub	2016-12-23 09:40:17.000000000 +0100
+++ binutils-2.28_cosmos/config.sub	2017-07-20 11:18:22.445631844 +0200
@@ -1407,7 +1407,7 @@
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
 	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* \
-	      | -onefs* | -tirtos* | -phoenix* | -fuchsia*)
+	      | -onefs* | -tirtos* | -phoenix* | -fuchsia* | -cosmos*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff -Nrdu binutils-2.28/gas/configure.tgt binutils-2.28_cosmos/gas/configure.tgt
--- binutils-2.28/gas/configure.tgt	2017-03-02 09:23:53.000000000 +0100
+++ binutils-2.28_cosmos/gas/configure.tgt	2017-07-20 11:18:54.425450781 +0200
@@ -137,6 +137,7 @@
 
   arm-*-aout)				fmt=aout ;;
   arm-*-coff)				fmt=coff ;;
+  arm-*-cosmos*)			fmt=elf ;;
   arm-*-phoenix*)			fmt=elf ;;
   arm-*-rtems*)				fmt=elf ;;
   arm-*-elf)				fmt=elf ;;
diff -Nrdu binutils-2.28/ld/configure.tgt binutils-2.28_cosmos/ld/configure.tgt
--- binutils-2.28/ld/configure.tgt	2017-03-02 09:23:54.000000000 +0100
+++ binutils-2.28_cosmos/ld/configure.tgt	2017-07-20 11:20:05.501081973 +0200
@@ -105,6 +105,7 @@
 			targ_extra_emuls="armelf armelf_nbsd armelfb_nbsd"  ;;
 arm-*-nto*)		targ_emul=armnto ;;
 arm-*-openbsd*)		targ_emul=armnbsd ;;
+arm-*-cosmos*)		targ_emul=armelf ;;
 arm-*-phoenix*)		targ_emul=armelf ;;
 arm-*-rtems*)		targ_emul=armelf ;;
 armeb-*-elf | armeb-*-eabi*)
diff -Nrdu binutils-2.28/ld/emulparams/armelf_cosmos.sh binutils-2.28_cosmos/ld/emulparams/armelf_cosmos.sh
--- binutils-2.28/ld/emulparams/armelf_cosmos.sh	1970-01-01 01:00:00.000000000 +0100
+++ binutils-2.28_cosmos/ld/emulparams/armelf_cosmos.sh	2017-07-20 11:23:07.128315224 +0200
@@ -0,0 +1,24 @@
+ARCH=arm
+SCRIPT_NAME=elf
+OUTPUT_FORMAT="elf32-littlearm"
+BIG_OUTPUT_FORMAT="elf32-bigarm"
+LITTLE_OUTPUT_FORMAT="elf32-littlearm"
+MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
+COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
+TEMPLATE_NAME=elf32
+EXTRA_EM_FILE=armelf
+GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
+
+DATA_START_SYMBOLS='PROVIDE (__data_start = .);';
+OTHER_TEXT_SECTIONS='*(.glue_7t) *(.glue_7) *(.vfp11_veneer) *(.v4_bx)'
+OTHER_BSS_SYMBOLS='__bss_start__ = .;'
+OTHER_BSS_END_SYMBOLS='_bss_end__ = . ; __bss_end__ = . ;'
+OTHER_END_SYMBOLS='__end__ = . ;'
+OTHER_SECTIONS='.note.gnu.arm.ident 0 : { KEEP (*(.note.gnu.arm.ident)) }'
+
+TEXT_START_ADDR=0x00010000
+TARGET2_TYPE=got-rel
+
+# ARM does not support .s* sections.
+NO_SMALL_DATA=yes 
diff -Nrdu binutils-2.28/ld/Makefile.am binutils-2.28_cosmos/ld/Makefile.am
--- binutils-2.28/ld/Makefile.am	2017-03-02 09:23:54.000000000 +0100
+++ binutils-2.28_cosmos/ld/Makefile.am	2017-07-20 11:26:10.147735937 +0200
@@ -184,6 +184,7 @@
 	earmelf_linux_eabi.c \
 	earmelf_nacl.c \
 	earmelf_nbsd.c \
+	earmelf_cosmos.c \
 	earmelf_phoenix.c \
 	earmelf_vxworks.c \
 	earmelfb.c \
@@ -790,6 +791,10 @@
   $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
   $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 
+earmelf_cosmos.c: $(srcdir)/emulparams/armelf_cosmos.sh \
+  $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
+  $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+
 earmelf_phoenix.c: $(srcdir)/emulparams/armelf_phoenix.sh \
   $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
   $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
diff -Nrdu binutils-2.28/ld/Makefile.in binutils-2.28_cosmos/ld/Makefile.in
--- binutils-2.28/ld/Makefile.in	2017-03-02 09:23:54.000000000 +0100
+++ binutils-2.28_cosmos/ld/Makefile.in	2017-07-20 11:27:52.831473360 +0200
@@ -553,6 +553,7 @@
 	earmelf_linux_eabi.c \
 	earmelf_nacl.c \
 	earmelf_nbsd.c \
+	earmelf_cosmos.c \
 	earmelf_phoenix.c \
 	earmelf_vxworks.c \
 	earmelfb.c \
@@ -1194,6 +1195,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_linux_eabi.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_nacl.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_nbsd.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_cosmos.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_phoenix.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_vxworks.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelfb.Po@am__quote@
@@ -2383,6 +2385,10 @@
   $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
   $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 
+earmelf_cosmos.c: $(srcdir)/emulparams/armelf_cosmos.sh \
+  $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
+  $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+
 earmelf_phoenix.c: $(srcdir)/emulparams/armelf_phoenix.sh \
   $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
   $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
