diff -Nrdu gcc-7.1.0/config.sub gcc-7.1.0_cosmos/config.sub
--- gcc-7.1.0/config.sub	2016-12-29 12:04:53.000000000 +0100
+++ gcc-7.1.0_cosmos/config.sub	2017-07-20 11:56:19.725454106 +0200
@@ -1409,7 +1409,7 @@
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
 	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* \
-	      | -onefs* | -tirtos* | -phoenix* | -fuchsia* | -redox*)
+	      | -onefs* | -tirtos* | -phoenix* | -fuchsia* | -redox* | -cosmos*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff -Nrdu gcc-7.1.0/gcc/config/arm/t-cosmos gcc-7.1.0_cosmos/gcc/config/arm/t-cosmos
--- gcc-7.1.0/gcc/config/arm/t-cosmos	1970-01-01 01:00:00.000000000 +0100
+++ gcc-7.1.0_cosmos/gcc/config/arm/t-cosmos	2017-07-20 11:57:44.369390390 +0200
@@ -0,0 +1,29 @@
+# Copyright (C) 1998-2017 Free Software Foundation, Inc.
+#
+# This file is part of GCC.
+#
+# GCC is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3, or (at your option)
+# any later version.
+#
+# GCC is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with GCC; see the file COPYING3.  If not see
+# <http://www.gnu.org/licenses/>.
+
+MULTILIB_OPTIONS     = marm/mthumb
+MULTILIB_DIRNAMES    = arm thumb
+MULTILIB_EXCEPTIONS  =
+MULTILIB_MATCHES     =
+
+MULTILIB_OPTIONS     += mfloat-abi=hard
+MULTILIB_DIRNAMES    += fpu
+MULTILIB_MATCHES     += mfloat-abi?hard=mhard-float
+
+MULTILIB_OPTIONS     += mno-thumb-interwork/mthumb-interwork
+MULTILIB_DIRNAMES    += normal interwork
diff -Nrdu gcc-7.1.0/gcc/config/cosmos.h gcc-7.1.0_cosmos/gcc/config/cosmos.h
--- gcc-7.1.0/gcc/config/cosmos.h	1970-01-01 01:00:00.000000000 +0100
+++ gcc-7.1.0_cosmos/gcc/config/cosmos.h	2017-07-20 11:59:19.041319963 +0200
@@ -0,0 +1,32 @@
+/* Copyright (C) 1998-2017 Free Software Foundation, Inc.
+
+This file is part of GCC.
+
+GCC is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3, or (at your option)
+any later version.
+
+GCC is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with GCC; see the file COPYING3.  If not see
+<http://www.gnu.org/licenses/>. */
+
+#undef TARGET_OS_CPP_BUILTINS
+#define TARGET_OS_CPP_BUILTINS()        \
+    do {                                \
+      builtin_define_std ("cosmos");    \
+      builtin_define_std ("unix");      \
+      builtin_assert ("system=cosmos"); \
+      builtin_assert ("system=unix");   \
+    } while(0);
+
+#define STD_LIB_SPEC "%{!shared:%{g*:-lg} %{!p:%{!pg:-lc}}%{p:-lc_p}%{pg:-lc_p}}"
+
+/* This will prevent selecting 'unsigned long int' instead of 'unsigned int' as 'uint32_t' in stdint-newlib.h. */
+#undef STDINT_LONG32
+#define STDINT_LONG32		0
diff -Nrdu gcc-7.1.0/gcc/config.gcc gcc-7.1.0_cosmos/gcc/config.gcc
--- gcc-7.1.0/gcc/config.gcc	2017-03-24 14:59:51.000000000 +0100
+++ gcc-7.1.0_cosmos/gcc/config.gcc	2017-07-20 12:01:10.721237822 +0200
@@ -818,6 +818,11 @@
       ;;
   esac
   ;;
+*-*-cosmos*)
+  gas=yes
+  gnu_ld=yes
+  default_use_cxa_atexit=yes
+  ;;
 *-*-phoenix*)
   gas=yes
   gnu_ld=yes
@@ -1121,6 +1126,12 @@
 	# The EABI requires the use of __cxa_atexit.
 	default_use_cxa_atexit=yes
 	;;
+arm*-*-cosmos*)
+	tm_file="elfos.h arm/unknown-elf.h arm/elf.h arm/bpabi.h"
+	tm_file="${tm_file} newlib-stdint.h cosmos.h"
+	tm_file="${tm_file} arm/aout.h arm/arm.h"
+	tmake_file="${tmake_file} arm/t-arm arm/t-bpabi arm/t-cosmos"
+	;;
 arm*-*-phoenix*)
 	tm_file="elfos.h arm/unknown-elf.h arm/elf.h arm/bpabi.h"
 	tm_file="${tm_file} newlib-stdint.h phoenix.h"
diff -Nrdu gcc-7.1.0/libgcc/config.host gcc-7.1.0_cosmos/libgcc/config.host
--- gcc-7.1.0/libgcc/config.host	2017-03-10 20:33:27.000000000 +0100
+++ gcc-7.1.0_cosmos/libgcc/config.host	2017-07-20 12:01:43.917213574 +0200
@@ -413,6 +413,13 @@
 	unwind_header=config/arm/unwind-arm.h
 	extra_parts="$extra_parts crti.o crtn.o"
 	;;
+arm*-*-cosmos*)
+	tmake_file="t-hardfp t-softfp arm/t-arm arm/t-elf arm/t-softfp arm/t-cosmos"
+	tmake_file="${tmake_file} arm/t-bpabi"
+	tm_file="$tm_file arm/bpabi-lib.h"
+	extra_parts="crtbegin.o crtend.o crti.o crtn.o"
+	unwind_header=config/arm/unwind-arm.h
+	;;
 arm*-*-phoenix*)
 	tmake_file="t-hardfp t-softfp arm/t-arm arm/t-elf arm/t-softfp arm/t-phoenix"
 	tmake_file="${tmake_file} arm/t-bpabi"

